### PR TITLE
Omarkj add ssl proxy

### PR DESCRIPTION
**Pull request by omarkj**
This PR contains considerable refactoring of the ranch proxy protocol library to
allow for code-reuse when implementing SSL support which, interestingly enough,
this PR also contains.

The SSL implementation does the proxy protocol handshake before upgrading the
socket to a SSL socket.

### CODE REVIEW

File: src/ranch_proxy_socket.erl [hunk 1]
```diff
@@ -0,0 +1,3 @@
+-module(ranch_proxy_socket).
+
+
```
Comment (ferd) [inline src/ranch_proxy_socket.erl]: What's this module for?

File: src/ranch_proxy_ssl.erl [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      Key == psk_identity;
+      Key == srp_identity;
+      Key == ssl_imp;
+      Key == reuse_sessions;
+      Key == reuse_session;
+      Key == next_protocol_advertised;
+      Key == client_preferred_next_protocols;
+      Key == client;
+      Key == log_alert;
+      Key == server_name_indication ->
```
Comment (ferd) [inline src/ranch_proxy_ssl.erl]: We're gonna need the options in https://github.com/ninenines/ranch/compare/master...heroku:sni

File: src/ranch_proxy_socket.erl (hunk 1 above)
Comment (omarkj) [inline src/ranch_proxy_socket.erl]: It's for deleting.

**Comment by ferd** [comment]
At a first glance this looks pretty sweet.

**Comment by jkakar** [comment]
From what I understand of this, it looks great, +1!

**Comment by omarkj** [comment]
I'm pulling this into master.

**Pull request merged**